File: code/python/289.py
class Solution:
    def gameOfLife(self, board: List[List[int]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """
        changed = []
        row, col = len(board), len(board[0])

        def live_or_die(x, y):
            live_neighors = 0
            for d in [[-1, -1], [-1, 0], [-1, 1], [0, -1], [0, 1], [1, -1], [1, 0], [1, 1]]:
                if 0 <= x + d[0] < row and 0 <= y + d[1] < col:
                    live_neighors += board[x + d[0]][y + d[1]]
            # 四条规则
            if board[x][y] == 1:
                if live_neighors < 2:
                    return False
                if live_neighors > 3:
                    return False
                return True
            else:
                if live_neighors == 3:
                    return False
                return True

        for i in range(row):
            for j in range(col):
                if not live_or_die(i, j):
                    changed.append((i, j))

        for i, j in changed:
            board[i][j] = 1 - board[i][j]

File: code/java/289.java
class Solution {
    public int checkBox(int[][] board,int i,int j,int m,int n){
        int result = 0;
        if(i>0&&j>0)
            result+=board[i-1][j-1]>0?1:0;
        if(i>0)
            result+=board[i-1][j]>0?1:0;
        if(i>0&&j<n-1)
            result+=board[i-1][j+1]>0?1:0;
        if(j>0)
            result+=board[i][j-1]>0?1:0;
        if(j<n-1)
            result+=board[i][j+1]>0?1:0;
        if(i<m-1&&j>0)
            result+=board[i+1][j-1]>0?1:0;
        if(i<m-1)
            result+=board[i+1][j]>0?1:0;
        if(i<m-1&&j<n-1)
            result+=board[i+1][j+1]>0?1:0;
        return result;
    }

    public void gameOfLife(int[][] board) {
        int m = board.length;
        int n = board[0].length;
        for(int i=0;i<m;i++){
            for(int j=0;j<n;j++){
                int temp = checkBox(board,i,j,m,n);
                if(board[i][j]==0)
                    board[i][j] = -1*temp;
                else{
                    if(temp==0)
                        temp = 1;
                    board[i][j] = temp;
                }
            }
        }
        for(int i=0;i<m;i++){
            for(int j=0;j<n;j++){
                if(board[i][j]!=-3&&board[i][j]!=2&&board[i][j]!=3)
                    board[i][j]=0;
                else
                    board[i][j]=1;
            }
        }
    }
}